### PR TITLE
JsonNumber implementation and tests

### PR DESCRIPTION
**Pull request by kasiabulat**
I implemented `JsonNumber` methods using string to keep the actual value (and added more tests for it). I also added more equals-related methods to JsonNumber, JsonBoolean and JsonString.

cc: @joperezr @bartonjs @ericstj @ahsonkhan @terrajobst

### CODE REVIEW

File: src/System.Text.Json/ref/System.Text.Json.cs [hunk 1]
```diff
@@ -90,14 +90,17 @@ public partial class JsonArray : System.Text.Json.JsonNode, System.Collections.G
         void System.Collections.Generic.ICollection<System.Text.Json.JsonNode>.CopyTo(System.Text.Json.JsonNode[] array, int arrayIndex) { }
         System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw null; }
     }
-    public partial class JsonBoolean : System.Text.Json.JsonNode
+    public partial class JsonBoolean : System.Text.Json.JsonNode, System.IEquatable<System.Text.Json.JsonBoolean>
```
Comment (joperezr) [inline src/System.Text.Json/ref/System.Text.Json.cs:97]: I suppose these you changed only because you were already changing JsonNumber. For the main fork PR only send JsonNumber class, which I suppose is already your plan, just wanted to reiterate :)

File: src/System.Text.Json/ref/System.Text.Json.cs [hunk 2]
```diff
@@ -264,7 +267,7 @@ public abstract partial class JsonNode
         public static System.Text.Json.JsonNode Parse(string json) { throw null; }
         public static bool TryGetNode(System.Text.Json.JsonElement jsonElement, out System.Text.Json.JsonNode jsonNode) { throw null; }
     }
-    public partial class JsonNumber : System.Text.Json.JsonNode
+    public partial class JsonNumber : System.Text.Json.JsonNode, System.IEquatable<System.Text.Json.JsonNumber>
```
Comment (joperezr) [inline src/System.Text.Json/ref/System.Text.Json.cs:274]: @bartonjs would it make sense to also make this implement IEquatable<int>, IEquatable<double>, etc.?

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs [hunk 3]
```diff
@@ -0,0 +1,218 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+
+namespace System.Text.Json
+{
+//#pragma warning disable xUnit1000
```
Comment (joperezr) [inline src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: NIT: now that you made the class public, just remove these comments.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 4]
```diff
@@ -7,89 +7,357 @@
...
+        }
 
-        public string GetString() { throw null; }
-        public byte GetByte() { throw null; }
-        public int GetInt32() { throw null; }
-        public short GetInt16() { throw null; }
-        public long GetInt64() { throw null; }
-        public float GetSingle() { throw null; }
-        public double GetDouble() { throw null; }
+        public string GetString()
```
Comment (joperezr) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: I'm not 100% about this one, shouldn't this be instead an override to ToString method? I get that in the others it totally makes sense to have the Get in order to get the value out, I just don't know how will users percieve a JsonNumber.GetString method, it is not trivial knowing that the way we store the value is with a string.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 5]
```diff
@@ -7,89 +7,357 @@
...
+        { 
+            return _value; 
+        }
+        public byte GetByte()
+        {
+            return byte.Parse(_value);
```
Comment (joperezr) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: NIT: all these methods can be written instead as oneliners which I think makes the file look cleaner:
```c#
public byte GetByte() => byte.Parse(_value);
```

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 6]
```diff
@@ -7,89 +7,357 @@
...
+            return ulong.Parse(_value);
+        }
 
-        public bool TryGetByte(out byte value) { throw null; }
-        public bool TryGetInt32(out int value) { throw null; }
-        public bool TryGetInt16(out short value) { throw null; }
-        public bool TryGetInt64(out long value) { throw null; }
-        public bool TryGetSingle(out float value) { throw null; }
-        public bool TryGetDouble(out double value) { throw null; }
+        public bool TryGetByte(out byte value)
```
Comment (joperezr) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: same for these, you could do:
```c#
public bool TryGetByte(out byte value) => byte.TryParse(_value, out value);
```

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 7]
```diff
@@ -7,89 +7,357 @@
...
+        {
+            _value = value.ToString();
+        }
+        public void SetSingle(float value)
+        {
+            _value = value.ToString();
+        }
+        public void SetDouble(double value)
+        {
+            _value = value.ToString();
```
Comment (joperezr) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: question: will this roundtrip when using exponential notation?

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 8]
```diff
@@ -7,89 +7,357 @@
 
 namespace System.Text.Json
 {
-    public partial class JsonNumber : JsonNode
+    public partial class JsonNumber : JsonNode, IEquatable<JsonNumber>
```
Comment (joperezr) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:12]: Would it also make sense to implement IComperable<JsonNumber>?

File: src/System.Text.Json/ref/System.Text.Json.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Text.Json/ref/System.Text.Json.cs:274]: > make this implement IEquatable, IEquatable, etc.?

Was one of them meant to be generic?

Comment (bartonjs) [reply src/System.Text.Json/ref/System.Text.Json.cs:274]: Assuming that your IEquatables were generic over numeric value types:

Not really.

The equality for this is either very complicated, simple but maybe not useful.  JSON numbers are any legal value in `Q` (the set of rational (real) numbers).  If we're looking at this from a mathy approach then "3" "3e0", "30e-1", "3.0", "3.00", and the infinite (up to memory limits) other possible ways of being "3" are all equal.  I doubt the implementation is going to take that into account.  I expect that it's just "do the two implementations have the same string, internally?".

An `IEquatable<int>` implementation could be written that does TryParse on the field and matches against the input.  But is it useful?

As far as explicit guidelines, all we have are that immutable value types should be IEquatable of themselves, and you should consider being IEquatable of yourself if you override Object.Equals, and if you're IComparable of yourself you should be IEquatable of yourself.

We don't have one that says "don't be IEquatable of not-you", but instinctively (as a mathematician) that "do not" makes sense to me because the inverse relationship isn't true (int isn't equatable to JsonNumber, so JsonNumber shouldn't be equatable to int, by symmetry)

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 9]
```diff
@@ -7,89 +7,357 @@
...
+        {
+            try
+            {
+                value = GetByte();
+                return true;
+            } 
+           catch
+            {
+                value = 0;
+                return false;
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: Try methods shouldn't be written as catch/ignore/return-false.  Implementing a non-Try method in terms of the Try is legitimate, though (the question I thought you asked me in person earlier).

But in this case, as @joperezr pointed out, the ideal implementation is to just defer to byte.TryParse (et al).

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 10]
```diff
@@ -7,89 +7,357 @@
...
-        public JsonNumber(short value) { }
-        public JsonNumber(int value) { }
-        public JsonNumber(long value) { }
-        public JsonNumber(float value) { }
-        public JsonNumber(double value) { }
+        public JsonNumber(string value)
+        {
+            SetString(value);
+        }
+        public JsonNumber(byte value)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: nit: add spaces between methods

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 11]
```diff
@@ -7,89 +7,357 @@
...
+                return true;
+            }
+            catch
+            {
+                value = 0;
+                return false;
+            }
+        }
+
+        public void SetString(string value)
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: This probably needs a better name, like `SetFormattedValue`.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 7 above)
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: This probably wants a specific format (like G15).  Whatever Utf8JsonWriter does.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 12]
```diff
@@ -7,89 +7,357 @@
...
-        public JsonNumber(string value) { }
-        public JsonNumber(byte value) { }
-        public JsonNumber(short value) { }
-        public JsonNumber(int value) { }
-        public JsonNumber(long value) { }
-        public JsonNumber(float value) { }
-        public JsonNumber(double value) { }
+        public JsonNumber(string value)
+        {
+            SetString(value);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: Explicitly validate and throw for null?
Unless you are implicitly relying on `Encoding.UTF8.GetBytes` to do the validation for you.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 13]
```diff
@@ -7,89 +7,357 @@
 
 namespace System.Text.Json
 {
-    public partial class JsonNumber : JsonNode
+    public partial class JsonNumber : JsonNode, IEquatable<JsonNumber>
     {
+        private string _value;
+
         public JsonNumber() { }
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: The default ctor here is the only time that `_value` can be `null`. Consider removing it, or initializing `_value` to `"0"`.

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+            
+            bool success = jsonNumber.TryGetByte(out result);
+            Assert.True(success);
+            Assert.Equal(value, result);
+        }
+
+        [Fact]
+        public static void TestShort()
+        {
+            var jsonNumber = new JsonNumber();
```
Comment (bartonjs) [inline src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:58]: The fact that this type has a default ctor surprised me.  There should be a test that tests the behaviors of this ctor (what throws, what succeeds, etc).

Either it should be 0 (in which case all methods work) or it should not exist.

File: src/System.Text.Json/ref/System.Text.Json.cs (hunk 2 above)
Comment (joperezr) [reply src/System.Text.Json/ref/System.Text.Json.cs:274]: > Was one of them meant to be generic

Yes sorry, codeflow didn't like the fact that I didn't use code blocks to define my generics 😆

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 9 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: As a general rule, try not to use exceptions (try/catch) as control flow (especially a catch-all):
https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+            bool success = jsonNumber.TryGetInt16(out result);
+            Assert.True(success);
+            Assert.Equal(value, result);
+        }
+
+        [Fact]
+        public static void TestInt()
+        {
+            var jsonNumber = new JsonNumber();
+            int value = 17;
```
Comment (bartonjs) [inline src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: Missing tests:

* Things used the typed ctors
* Creating a value with one type, then changing it to another.
* When the Get methods fail / the TryGet methods return false.
* ToString/GetString/whatever should always succeed.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 16]
```diff
@@ -7,89 +7,357 @@
...
+
+        public override bool Equals(object obj)
+        {
+            return obj is JsonNumber number &&
+                   _value == number._value;
+        }
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(_value);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: nit: How about just `_value.GetHashCode()`?

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        [Theory]
+        [InlineData("3,14")]
+        [InlineData("this is not a number")]
+        [InlineData("NAN")]
+        [InlineData("0.")]
+        [InlineData("008")]
+        [InlineData("5e")]
+        public static void TestInvalidString(string value)
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:230]: We should consider expanding on the "invalid string" cases.

See these for some more:
https://github.com/dotnet/corefx/blob/9e994a5923d426f0e64ed3b8f9b93df6f214a42a/src/System.Text.Json/tests/Utf8JsonReaderTests.cs#L4382-L4393

Consider also including white space before/after.

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+            bool success = jsonNumber.TryGetUInt32(out result);
+            Assert.True(success);
+            Assert.Equal(value, result);
+        }
+
+        [Fact]
+        public static void TestUInt64()
+        {
+            var jsonNumber = new JsonNumber();
+            ulong value = 5;
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: Consider adding more, especially ones that are within the uint64 range but not in int32/int64 ranges (like uint64.maxvalue).

It would be good to test the boundary values.

Comment (ahsonkhan) [reply src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: Same goes for other overloads (including negative numbers).

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        [Fact]
+        public static void TestDouble()
+        {
+            var jsonNumber = new JsonNumber();
+            double value = 3.14;
+
+            jsonNumber.SetDouble(value);
+            double result = jsonNumber.GetDouble();
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: What do you get if you call `jsonNumber.GetInt32()` here? Is it valid to do so?

It might be good to structure the tests so that you can create a JsonNumber from some input (say a string) and then call all sorts of mismatching `GetX()` APIs on them.

So, SetX(some numeric type), GetX(another numeric type).

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+        [InlineData(3.14f)]
+        [InlineData(0x2A)]
+        [InlineData(0b_0110_1010)]
+        [InlineData("1e400")]
+        [InlineData("1e+100000002")]
+        [InlineData("184467440737095516150.184467440737095516150")]
+        public static void TestJsonNumberConstructor(object value)
+        {
+            // should not throw any exceptions:
+            var jsonNumber = new JsonNumber(value as dynamic);
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: I would consider adding to this test and showcasing the behavior of each {Try}GetX() API.
```suggestion
            var jsonNumber = new JsonNumber(value as dynamic);
            // call and assert jsonNumber.GetDouble();
            // call and assert jsonNumber.TryGetDouble();
            // call and assert jsonNumber.GetSingle();
            // ... etc.
```

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs (hunk 14 above)
Comment (ahsonkhan) [reply src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:58]: > The fact that this type has a default ctor surprised me.

Yep, same here.

**Comment by ahsonkhan** [comment]
cc @JamesNK

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 8 above)
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:12]: Presumably we are defering API documentation (`/// xml comments` like summary/parameters/returns/exceptions) until after the API is more stable. Is the plan to add them once the type/impl/tests moves to corefx?

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs (hunk 15 above)
Comment (ahsonkhan) [reply src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: Also implicit casts.

File: src/System.Text.Json/ref/System.Text.Json.cs [hunk 21]
```diff
@@ -283,6 +286,7 @@ public partial class JsonNumber : System.Text.Json.JsonNode
         [System.CLSCompliantAttribute(false)]
         public JsonNumber(ulong value) { }
```
Comment (ahsonkhan) [inline src/System.Text.Json/ref/System.Text.Json.cs:292]: What about decimal?

**Comment by JamesNK** [comment]
> cc @JamesNK

I don't have any context here. Could you point me at a spec/proposal/issue that discusses what these APIs are for

**Comment by ahsonkhan** [comment]
> I don't have any context here. Could you point me at a spec/proposal/issue that discusses what these APIs are for

This is still in the initial prototype/design phase and maybe @joperezr / @kasiabulat have some doc/spec they can point you to (if we don't have one yet, it might be good to write one up at some point).

We are starting to fill in implementation details and then go through the API review/design in corefx within the next month or so (before @kasiabulat's internship ends), so there is no issue yet.

Essentially, we are looking to create a modifiable, dictionary-backed Json DOM API to complement the readonly `JsonDocument`/`JsonElement`, with type hierarchy consisting of the base type `JsonNode` (rather than a single union type `JsonElement`).
The user should be able to build up a structured in-memory representation of the JSON payload and modify it. That includes, remove, add, and update (similar to Newtonsoft.Json's `JToken`).

The `WritableJsonApiTests.cs` contains a bunch of usage scenarios that we want to enable with these types. For example:
https://github.com/kasiabulat/corefx/blob/25109c364e5bcb2d70a2a7b91f620e28faab27ca/src/System.Text.Json/tests/WritableJsonApiTests.cs#L19-L42

I cc'd you primarily to get your eyes on it, and to get your insights early (if applicable), especially to avoid common pitfalls or design mistakes that you may have encountered (and to get your general opinion on the design). If you see any big rocks or have significant concerns, it would be great to discuss and address them before we have gone too deep with the current approach.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 4 above)
Comment (kasiabulat) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: Sure, I'll change it.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 6 above)
Comment (kasiabulat) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: Oh yes, thanks! :)

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 11 above)
Comment (kasiabulat) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: Ok, I'll change it.

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs (hunk 14 above)
Comment (kasiabulat) [reply src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:58]: Ok, I can change it to "0".

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs (hunk 15 above)
Comment (kasiabulat) [reply src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: Thanks, I'll add those.

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs (hunk 18 above)
Comment (kasiabulat) [reply src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: Ok, sure!

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 7 above)
Comment (kasiabulat) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: Ok, I checked it - "G15" is default for ToString and Utf8JsonWriter uses default StandardFormat - that is also "G", but with unspecified precision. Should I leave the default format or change it to "G17" or something else?

Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: I thought we had changed Utf8JsonWriter to specify a precision.  If not, nothing is needed here :smile:.

**Comment by kasiabulat** [comment]
I changed methods into one-liners and added more tests, but still need to add some for Get/TryGet for different types (correct and incorrect examples). Personally, I think it should be possible to GetInt after setting something that fits into int and not possible if it doesn't etc. And in current implementation I forget the information of what exact type was stored.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 22]
```diff
@@ -5,91 +5,161 @@
...
+
         [CLSCompliant(false)]
-        public static implicit operator JsonNumber(ulong value) { throw null; }
+        public bool TryGetUInt64(out ulong value) => ulong.TryParse(_value, out value);
+
+
+        public void SetFormattedValue(string value)
+        {
+            if (value == null)
+                throw new NullReferenceException("Expected number, but instead got null.");
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: This should be `ArgumentNullException`. I believe we never explicitly throw NullRef (and its only thrown by the runtime). For example:
https://github.com/dotnet/corefx/blob/9d6729a940952dcc14dc8683218684f4da29198a/src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs#L93-L96

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 23]
```diff
@@ -5,91 +5,161 @@
...
+            if (value == "")
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: nit: Checking length will likely be much faster (and avoid a call to `.Equals`). [sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgMwAEu5AwuQN7HlOUXAD2bANuQLJ4AUWPAAZyANwicAlPUbN5WAOzjJ5ALxryAIi0BuOUwC+BluXZdeOQSJXTZReQuUTOAOgAycAHYBzGAAt1TWF9B2ZjIkMgA===) 
```suggestion
            if (value.Length == 0)
```

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 24]
```diff
@@ -5,91 +5,161 @@
...
+                throw new ArgumentException("Expected number, but instead got empty string.");
+            
+            JsonWriterHelper.ValidateNumber(Encoding.UTF8.GetBytes(value).AsSpan());
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:108]: Consider adding "throw argumentexception" validation to a copy of `JsonWriterHelper.ValidateNumber` (with Debug.Asserts changed to actual checks).

The length check can go there as well:
https://github.com/dotnet/corefx/blob/9d6729a940952dcc14dc8683218684f4da29198a/src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs#L137-L148

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 25]
```diff
@@ -5,91 +5,161 @@
...
+            _value = value;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:109]: I would suspect these strings to generally be small (say < 256 characters). Consider instead, stackalloc a span and use the `Span<byte>` overload of `Encoding.UTF8.GetBytes` to avoid the allocation on .NET Core 3.0 (and use the arraypool for larger length strings).
https://github.com/dotnet/corefx/blob/ef5b086fc5f134e3244d0bca10d70db8055fa1c6/src/System.Runtime/ref/System.Runtime.cs#L7683

OR

Validate the value as chars.

Something to consider to optimize the implementation a bit (but that can be deferred to later, if you like).

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 26]
```diff
@@ -5,91 +5,161 @@
...
+        
+        public byte GetByte() => byte.Parse(_value);
+        
+        public short GetInt16() => short.Parse(_value);
+        
+        public int GetInt32() => int.Parse(_value);
+        
+        public long GetInt64() => long.Parse(_value);
+        
+        public float GetSingle() => float.Parse(_value);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:56]: Why is `decimal` not supported here? I think we should support all the .NET types the existing `JsonDocument`/`JsonElement` supports

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs [hunk 27]
```diff
@@ -5,91 +5,161 @@
...
+        
+        public double GetDouble() => double.Parse(_value);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:58]: **Question:** Would escaped characters be supported? Especially since the `Utf8JsonWriter` escapes the `+` character by default (since it uses `JavascriptEncoder.Default`).

```C#
Console.WriteLine(double.Parse("25.1e+1")); // valid - 251
Console.WriteLine(double.Parse("25.1e\\u002B1")); // Fails
```

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs [hunk 28]
```diff
@@ -0,0 +1,519 @@
...
+            Assert.Equal(145, result);
+        }
+
+        [Fact]
+        public static void TestUpcasts()
+        {
+            byte value = 17;
+            var jsonNumber = new JsonNumber(value);
+
+            // Geeting other types should also succeed:
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: nit: typo
```suggestion
            // Getting other types should also succeed:
```

Same elsewhere (everywhere else it was copied).

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 22 above)
Comment (kasiabulat) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: Ok, I'll change it.

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs [hunk 29]
```diff
@@ -0,0 +1,519 @@
...
+            short shortResult = jsonNumber.GetInt16();
+            Assert.Equal(value, shortResult);
+            bool success = jsonNumber.TryGetInt16(out shortResult);
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: nit: It doesn't look like the `success` local is used for anything meaningful.

How about just assert:
```C#
Assert.True(jsonNumber.TryGetInt16(out shortResult));
```

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 26 above)
Comment (kasiabulat) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:56]: Yes, I plan to add it.

File: src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs (hunk 29 above)
Comment (kasiabulat) [reply src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs]: Sure, thanks!

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 23 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: Or combine both with `string.IsNullOrEmpty(value)`

File: src/System.Text.Json/ref/System.Text.Json.cs (hunk 21 above)
Comment (kasiabulat) [reply src/System.Text.Json/ref/System.Text.Json.cs:292]: I'll add it.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 23 above)
Comment (kasiabulat) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs]: I split those cases to throw different exceptions.

File: src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs (hunk 8 above)
Comment (kasiabulat) [reply src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:12]: I'll add documentation in main corefx PR.

**Comment by kasiabulat** [comment]
I added decimal support and addressed other minor issues. There are still some questions left to discuss, but I am merging this PR and I will create one to the main corefx fork.

**Comment by kasiabulat** [comment]
I created the issue for my project: https://github.com/dotnet/corefx/issues/39922
